libexec/rtld-elf/rtld_c18n.c

```diff
@@ -0,0 +1,700 @@
+} comparts;
+
+static void
+comparts_data_expand(compart_id_t capacity)
+{
+	const struct compart **data;
+
+	data = realloc(comparts.data, sizeof(*data) * capacity);
+	if (data == NULL)
+		rtld_fatal("realloc failed");
+	comparts.data = data;
+	comparts.capacity = capacity;
+}
+
+static bool
+string_search(const char *const strs[], const char *sym)
+{
+	if (strs != NULL)
+		for (; *strs != NULL; ++strs)
+			if (strcmp(sym, *strs) == 0)
+				return (true);
+	return (false);
+}
+
+static const struct compart *const *
+compart_get_or_create(const char *lib)
+{
+	struct {
+		const char *libraries[2];
+		struct compart com;
+	} *buf;
+	for (compart_id_t i = 1; i < comparts.size; ++i)
+		if (string_search(comparts.data[i]->libraries, lib))
+			return (&comparts.data[i]);
+	if (comparts.size == comparts.capacity)
+		comparts_data_expand(comparts.capacity * 2);
+	buf = xmalloc(sizeof(*buf));
+	buf->libraries[0] = lib;
+	buf->libraries[1] = NULL;
+	buf->com = (struct compart) {
+		.name = lib,
+		.libraries = buf->libraries
+	};
+	comparts.data[comparts.size] = &buf->com;
+	return (&comparts.data[comparts.size++]);
+}
+
+compart_id_t
+compart_id_allocate(const char *name)
+{
+	const struct compart *const *com = compart_get_or_create(name);
+	return (com - comparts.data);
+}
+
+static bool
+tramp_should_include(const Obj_Entry *reqobj, const struct tramp_data *data)
+{
+	const char *sym;
+
+	if (data->target == NULL)
+		return (false);
+
+	if (reqobj == NULL)
+		return (true);
+
+	if (reqobj->compart_id == data->defobj->compart_id)
+		return (false);
+
+	sym = strtab_value(data->defobj, data->def->st_name);
+
+	if (string_base_search(&trusted_globals, sym) != -1)
+		return (false);
+
+	return (true);
+}
+
+/*
+ * Trampolines
+ */
+struct tramp_pg {
+	void *_Atomic cursor;		/* Start of unused space */
+	SLIST_ENTRY(tramp_pg) link;	/* Link to next page */
+	void *trampolines[];		/* Start of trampolines */
+};
+
+/*
+* 64K is the largest size such that any capability-aligned proper
+* sub-range can be exactly bounded by a Morello capability.
+*/
+#define C18N_TRAMPOLINE_PAGE_SIZE 64 * 1024
+
+static struct tramp_pg *
+tramp_pg_new(struct tramp_pg *next)
+{
+	struct tramp_pg *pg = mmap(NULL,
+	    C18N_TRAMPOLINE_PAGE_SIZE,
+	    PROT_READ | PROT_WRITE | PROT_EXEC,
+	    MAP_ANON | MAP_PRIVATE,
+	    -1, 0);
+	if (pg == MAP_FAILED)
+		rtld_fatal("mmap failed");
+	atomic_store_explicit(&pg->cursor, pg->trampolines,
+	    memory_order_relaxed);
+	SLIST_NEXT(pg, link) = next;
+	return (pg);
+}
+
+static void *
+tramp_pg_push(struct tramp_pg *pg, size_t len)
+{
+	void *cur = atomic_load_explicit(&pg->cursor, memory_order_relaxed);
+	void *n_cur, *tramp;
+	do {
+		ptraddr_t n_cur_addr;
+		ptrdiff_t n_cur_len;
+
+		tramp = cheri_setbounds(cur, len);
+		n_cur_addr = roundup2(cheri_gettop(tramp), _Alignof(void *));
+		n_cur_len = cheri_gettop(cur) - n_cur_addr;
+
+		n_cur = cheri_setaddress(cur, n_cur_addr);
+		n_cur = cheri_setboundsexact(n_cur, n_cur_len);
```

Comment:
This is really lazy programming and an ugly interface. Track what's left and return NULL if there's not enough space.